src/main/java/com/google/javascript/gents/ModuleConversionPass.java

```diff
@@ -280,8 +294,109 @@ public boolean shouldTraverse(NodeTraversal t, Node n, Node parent) {
     }
   }
 
+  /** Encapsulates a {@code goog.require(...)} statement. */
+  class ModuleImport {
+    /** require statement Node. */
+    private Node originalImportNode;
+
+    /**
+     * LHS of the requirement statement. If the require statement has no LHS, then local name is the
+     * suffix of the required namespace/module.
+     */
+    private List<String> localNames;
+
+    /**
+     * Full local name is used for rewriting imported variables in the file later on. There's a one
+     * to one relationship between fullLocalNames and localNames.
+     */
+    private List<String> fullLocalNames;
+
+    /**
+     * For {@code goog.require(...)} with no LHS and no side effects, then we use the required
+     * namespace/module's suffix as the local name. The backup name is useful in this case to avoid
+     * conflicts when the required namespace/module's suffix is the same with a named exported from
+     * the same file.
+     */
+    private String backupName;
+
+    /** The required namespace or module name */
+    private String requiredNamespace;
+
+    /**
+     * {@code true}, if the original require statement contains destructuring imports. For
+     * destructuring imports, there are one or more local names and full local names. For non
+     * destructuring imports, there is exactly one local name and one full local name.
+     */
+    private boolean isDestructuringImport;
+
+    /** FileModule for the imported file */
+    private FileModule module;
+
+    /** Referenced file */
+    private String referencedFile;
+
+    /**
+     * The last part of the required namespace and module, used to detect local name conflicts and
+     * calculate the backup name
+     */
+    private String moduleSuffix;
+
+    private ModuleImport(
+        Node originalImportNode,
+        List<String> fullLocalNames,
+        String requiredNamespace,
+        boolean isDestructuringImport) {
+      this.originalImportNode = originalImportNode;
+      this.fullLocalNames = fullLocalNames;
+      this.requiredNamespace = requiredNamespace;
+      this.isDestructuringImport = isDestructuringImport;
+      this.module = namespaceToModule.get(requiredNamespace);
+      this.referencedFile =
+          pathUtil.getImportPath(originalImportNode.getSourceFileName(), module.file);
+      this.moduleSuffix = nameUtil.lastStepOfName(requiredNamespace);
+      this.backupName = this.moduleSuffix;
```

Comment:
Not sure if possible, but might be a good idea to encapsulate the backup logic within ModuleImport class.

Is it possible for the decision whether to use localName or backupName to be behind a getNameFor...() method?

Reply:
That's a good idea, but I tried it out and it seems to complicate the logic.
backup name is only used in a very specific case: A non destructuring require that converts to be two import statements, both `import {value} from './file';` and `import * as valueExports from './file';`. In this specific case, the latter import will use the backup name. So we need to add some extra flags in ModuleImport class for getNameFor...() method to work properly.
When it's inlined in the `convertNonDestructuringRequireToImportStatements()`, it's quite clear when we switch to backup name. So I'd keep it as is.